Fix #7378: Warn if tank (mixed or stratified) falls below 2C to indicate potential freeze

jmarrec opened the pull request:
Pull request overview
---------------------

Fix #7378: Warn if tank (mixed or stratified) falls below 2C to indicate potential freeze

### Work Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] At least one of the following appropriate labels must be added to this PR to be consumed into the changelog:
   - Defect: This pull request repairs a github defect issue.  The github issue should be referenced in the PR description
   - Refactoring: This pull request includes code changes that don't change the functionality of the program, just perform refactoring
   - NewFeature: This pull request includes code to add a new feature to EnergyPlus
   - Performance: This pull request includes code changes that are directed at improving the runtime performance of EnergyPlus
   - DoNoPublish: This pull request includes changes that shouldn't be included in the changelog

### Review Checklist
This will not be exhaustively relevant to every PR.
 - [ ] Functional code review (it has to work!)
 - [ ] If defect, results of running current develop vs this branch should exhibit the fix
 - [ ] CI status: all green or justified
 - [ ] Performance: CI Linux results include performance check
 - [ ] Unit Test(s)
 - C++ checks:
   - [ ] Argument types
   - [ ] If any virtual classes, ensure virtual destructor included, other things
 - IDD changes:
   - [ ] Verify naming conventions and styles, memos and notes and defaults
   - [ ] Open windows IDF Editor with modified IDD to check for errors
   - [ ] If transition, add to input rules file for interfaces
   - [ ] If transition, add transition source
   - [ ] If transition, update idfs
 - [ ] If new idf included, locally check the err file and other outputs
 - [ ] Required documentation updates?
 - [ ] ExpandObjects changes?
 - [ ] If output changes, including tabular output structure, add to output rules file for interfaces

jmarrec committed:
Add unit test for #7378

jmarrec committed:
Fix #7378: Warn if tank (mixed or stratified) falls below 2C to indicate potential freeze

jmarrec committed:
Merge branch 'develop' into PR_opened/Fix_7378_StorageTankTempLimit

Myoldmopar commented:
@jmarrec there are still 2 unit tests failing here.  Please fix up and we'll work on reviewing this.  Thanks!

jmarrec committed:
Merge remote-tracking branch 'upstream/develop' into PR_opened/Fix_7378_StorageTankTempLimit

jmarrec committed:
Init FreezingErrorIndex to zero in WaterThermalTankData Ctor. And correctly use that one in ShowRecurringWarningErrorAtEnd

jmarrec commented:
@Myoldmopar Sorry about this. I had made a couple of mistake in the use of the int FreezingErrorIndex (forgot to initialize it to zero, and was using another one in ShowRecurringWarningErrorAtEnd due to silly copy paste), and I probably  forgot to run all tests so I wasn't seeing any failures locally...

Should be good to go now, I have 100% success locally. Will let CI test it

jmarrec commented:
Two err diffs - `HeatRecoveryPlantLoopAuto` and `MicroCogeneration` - respectively 60 times and 30 times, only during warmup.

I think I should wrap the blocks I added into a `if (!WarmupFlag)` to only report during simulation. (Feel free to revert https://github.com/NREL/EnergyPlus/pull/7395/commits/8ec2c2deb664433d904baa3f391341655025bc14 if reporting during warmup is warranted)

jmarrec committed:
Only report if not during warmup.

rraustad commented:
@jmarrec defect file shows only recurring warnings. Typically a warning will be issued at the first occurrence of the problem with a time stamp, then recurring warning count takes over.

```
   *************  ** Warning ** WaterHeater:Mixed = 'WATER HEATER':  Temperature of tank < 2C indicates of possibility of freeze
   *************  **   ~~~   **   This error occurred 20255 total times;
   *************  **   ~~~   **   during Warmup 656 times;
   *************  **   ~~~   **   during Sizing 0 times.
   *************  **   ~~~   **   Max=1.99991 {C}  Min=-19.464834 {C}
```

@mjwitte comment?

rraustad commented:
Here's an example where info is provided specific to the problem:

```
   ** Warning ** CoilSystem:Cooling:DX - Iteration limit exceeded calculating DX unit sensible part-load ratio for unit = GUID26_SZVAV_ZN_ROOM_00_03SCC-1SYSTEM
   **   ~~~   ** Estimated part-load ratio  = 1.000
   **   ~~~   ** Calculated part-load ratio = 1.000
   **   ~~~   ** The calculated part-load ratio will be used and the simulation continues. Occurrence info:
   **   ~~~   **  Environment=RUNPERIOD 1, at Simulation time=05/30 12:40 - 13:00
```

```
   *************  ** Warning ** CoilSystem:Cooling:DX "GUID26_SZVAV_ZN_ROOM_00_03SCC-1SYSTEM" - Iteration limit exceeded calculating sensible part-load ratio error continues. Sensible PLR statistics follow.
   *************  **   ~~~   **   This error occurred 1 total times;
   *************  **   ~~~   **   during Warmup 0 times;
   *************  **   ~~~   **   during Sizing 0 times.
   *************  **   ~~~   **   Max=1.  Min=1.
```

jmarrec commented:
I'll take a look. Before i wrapped in if(!WarmupFlag) it was definitely producing both.

EDIT/note: I think you aren't looking with the latest commit (https://github.com/NREL/EnergyPlus/pull/7395/commits/8ec2c2deb664433d904baa3f391341655025bc14) btw, since now it shouldn't report for Warmup. Perhaps you reverted that commit on purpose though.

jmarrec commented:
@rraustad I added (and cleaned invalid stuff, and transitioned to 9.2.0) the defect files added by user on https://github.com/NREL/EnergyPlus/issues/7378#issuecomment-512287199 at https://github.com/NREL/EnergyPlusDevSupport/tree/master/DefectFiles/7000s/7378

I am pasting the relevant section of eplusout.err which will showcase that there is a single warning issued, then a recurring error summary. I've also created a gist with the full eplusout.err log for context (see [here](https://gist.github.com/jmarrec/efe8f509988d22b56629e45b6cf1b18c))


----

## Relevant sections of eplusout.err

### Regular warning (once)
https://gist.github.com/jmarrec/efe8f509988d22b56629e45b6cf1b18c#file-eplusout-err-L43:L44

```
   ** Warning ** CalcWaterThermalTankMixed: WaterHeater:Mixed = 'WATER HEATER':  Temperature of tank < 2C indicates of possibility of freeze. Tank Temperature = 1.28 C.
   **   ~~~   **  Environment=RUN PERIOD 1, at Simulation time=01/02 04:00 - 05:00
```

### Inside Recurring Error Summary:

```
   *************
   ************* ===== Recurring Error Summary =====
   ************* The following recurring error messages occurred.
   *************
```

https://gist.github.com/jmarrec/efe8f509988d22b56629e45b6cf1b18c#file-eplusout-err-L110:L114

```
   *************  ** Warning ** WaterHeater:Mixed = 'WATER HEATER':  Temperature of tank < 2C indicates of possibility of freeze
   *************  **   ~~~   **   This error occurred 19599 total times;
   *************  **   ~~~   **   during Warmup 0 times;
   *************  **   ~~~   **   during Sizing 0 times.
   *************  **   ~~~   **   Max=1.99991 {C}  Min=-19.464834 {C}
```

rraustad commented:
That looks good. This was a new branch on my end so why I didn't get the latest code is a mystery to me. I'll get this in shortly.

rraustad commented:
Confirmed initial warning at 1/2 4:00-5:00 and recurring warnings match what is shown above. CI results are clean and green.

Myoldmopar commented on src/EnergyPlus/WaterThermalTanks.cc:
```diff
@@ -7547,6 +7547,25 @@ namespace WaterThermalTanks {
+        if (!WarmupFlag) {
+            // Warn for potential freezing when avg of final temp over all nodes is below 2°C (nearing 0°C)
+            if (Tank.TankTemp < 2) {
+                if (Tank.FreezingErrorIndex == 0) {
```
Show error the first time, then recur.  Yep, this looks good.

Myoldmopar commented on src/EnergyPlus/WaterThermalTanks.cc:
```diff
@@ -7547,6 +7547,25 @@ namespace WaterThermalTanks {
         Tank.TankTemp = TankTemp;            // Final tank temperature for carry-over to next timestep
         Tank.TankTempAvg = TankTempAvg;      // Average tank temperature over the timestep for reporting
+
+        if (!WarmupFlag) {
```
Protecting here is fine.

Myoldmopar commented on tst/EnergyPlus/unit/WaterThermalTanks.unit.cc:
```diff
@@ -3125,3 +3125,214 @@ TEST_F(EnergyPlusFixture, Desuperheater_Multispeed_Coil_Test)
+    });
+    EXPECT_TRUE(compare_err_stream(error_string, true));
+
+}
```
Great unit tests.

Myoldmopar reviewed (approved):
I'll build this locally with latest develop and make sure unit test suite runs and this should be able to drop in.

Myoldmopar commented:
Windows never finished this up because I took the machine down yesterday during what I thought was a gap between runs.  I'm not worried about waiting on it.  As soon as my local tests are done I'll merge and wrap this up.

Myoldmopar commented:
...And they just finished successfully.  Merging.
